services/terrain/src/terrain/services/permanent_id_requests.clj

```diff
@@ -38,11 +44,21 @@
              :path staging-dest})))
   data-item)
 
+(defn- validate-publish-dest
+  [{:keys [path] :as data-item}]
+  (let [publish-dest (ft/path-join (config/permanent-id-publish-dir) (ft/basename path))]
+    (when (data-info/path-exists? (config/irods-user) publish-dest)
+      (throw+ {:type :clojure-commons.exception/exists
+               :error "A folder with this name has already been published."
+               :path publish-dest})))
```

Comment:
(long response incoming; I think what I'm saying isn't coming through concisely. sorry about that!)

I understood what it's doing as far as the validation; none of this is happening at the same time as things actually moving around to all of the paths being validated in iRODS. I think that it's reasonable to have it where it is for the sake of the more specific message even if it doesn't do anything else (like being a different code), but this seemed like a place to have a discussion about the boundaries and roles of the microservices -- whether data-info is only tasked with providing information in cases like this, and thus only taking action or throwing errors when it comes to modifying things in iRODS (i.e.: get the info from data-info, throw the error in terrain) or if its role also can include validating things in iRODS, like a file existing or not. Another version of this might say that data-info's scope includes validation as a primary task, even for operations being performed elsewhere, but that it can only throw generic errors, such that this bit of code might call a validator endpoint and catch an ERR_EXISTS (if it gets one) to rethrow as a more descriptive error.

For my own part, parallel to this work, I'm trying to figure out what to do with the urlupload endpoint -- it needs to validate some very similar things to this (user exists in irods, destination directory writeable, destination file doesn't already exist). Since we don't want jargon in terrain, of course this will require some calls to data-info, but it's not clear which calls exactly: a call each to existence-marker and to stat-gatherer, with terrain needing to parse the permissions and existence information out and throw its own errors; a call or several to endpoints which check things and either return success responses or throw errors; or just one call and let data-info call out to apps itself to start the job. Past discussions have largely if not completely ruled out the third (it could still make sense to have data-info call out to apps, if we consider the fact we're using an app to do urlupload as an implementation detail for something which is fundamentally an action in iRODS i.e. belongs in data-info), but which of the first/second is ideal isn't clear to me. What you're advocating would be the first, which is possibly easier to explain but more code and more requests into data-info, and what I've been advocating is the second, which seems to be harder to explain (unless that's just me being bad at explaining things, which is possible :P) but reduces the cross-talk between the services that's needed as well as the need for terrain to parse information out of data-info requests solely for validation. Your case the number of calls to data-info is probably the same either way, but it could still be a cleaner separation of concerns or less code one way or the other.

Anyway, how it is is fine, since the point here was the discussion. But I'm still curious to thresh out these boundaries better.